# Santa Fe Benchmark, PM

In [1]:
from pycgp.benchmarks.santafe import PARAMS, EV_PARAMS, X, santafe_cost_function
from pycgp.evolution import evolution
import numpy as np
import random

In [2]:
def run_experiment(params, ev_params, x):
    rstat = []
    for i in range(0, 20):
        print(i, end=', ')

        result = evolution(PARAMS, EV_PARAMS, X, None)

        rstat.append([santafe_cost_function(None, individual.execute(X)) for individual in result['final']])
    
    #print('Best fitness: {}'.format(np.min(stats)))
    #print('mean and std of fitness of last generation: {}, {}'.format(np.mean(stats), np.std(stats)))
    #print('Mean and std of best fitness: {}, {}'.format(np.mean(np.min(stats, axis=1)), np.std(np.min(stats, axis=1))))
    # best fitness, mean of last generation, std of last generation, mean of best individual, std of best individual
    results = [
        np.min(rstat), np.mean(rstat), np.std(rstat), np.mean(np.min(rstat, axis=1)), np.std(np.min(rstat, axis=1)) 
    ]
    print(results)
    return results
    

## Point mutation

In [3]:
from pycgp.mutation import point_mutation
all_measurements = {}

### PM, 10 nodes

In [4]:
%%time
random.seed(1)

PARAMS['n_cols']      = 10
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = False

all_measurements['10,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -23.399999999999999, 18.558017135459277, -36.399999999999999, 18.575252353602092]
CPU times: user 1min 47s, sys: 3.8 ms, total: 1min 47s
Wall time: 1min 47s


### PM, 50 nodes

In [5]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']      = 50
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = False

all_measurements['50,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -47.32, 28.033865234747779, -53.950000000000003, 26.736632173854659]
CPU times: user 5min 34s, sys: 20 ms, total: 5min 34s
Wall time: 5min 34s


### PM, 100 nodes

In [6]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']      = 100
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = False

all_measurements['100,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -56.409999999999997, 25.786079577942822, -63.149999999999999, 23.86687034363743]
CPU times: user 18min 26s, sys: 144 ms, total: 18min 26s
Wall time: 18min 28s


### PM, 10 nodes, Gems

In [7]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']      = 10
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = True

all_measurements['10,false,5'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -21.73, 24.185059437594937, -33.5, 28.355775425828156]
CPU times: user 2min 29s, sys: 24 ms, total: 2min 29s
Wall time: 2min 30s


### PM, 50 nodes, Gems

In [8]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']      = 50
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = True

all_measurements['50,false,5'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -54.25, 31.271512595331874, -61.850000000000001, 28.009418058931534]
CPU times: user 8min 38s, sys: 108 ms, total: 8min 38s
Wall time: 8min 41s


### PM, 100 nodes, Gems

In [9]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']      = 100
EV_PARAMS['mutation'] = point_mutation
EV_PARAMS['gems']     = True

all_measurements['100,false,5'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -52.43, 27.603353057192162, -56.350000000000001, 25.737666949434249]
CPU times: user 28min 16s, sys: 468 ms, total: 28min 17s
Wall time: 28min 21s


## jBox size 10

In [10]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 10
EV_PARAMS['mutation']   = point_mutation
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['10,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -19.93, 20.559306894932035, -32.149999999999999, 24.413674446916016]
CPU times: user 3min 23s, sys: 140 ms, total: 3min 23s
Wall time: 3min 26s


In [11]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 50
EV_PARAMS['mutation']   = point_mutation
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['50,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -43.210000000000001, 30.966528704393067, -48.450000000000003, 29.537222279693125]
CPU times: user 5min 54s, sys: 72 ms, total: 5min 55s
Wall time: 5min 55s


In [12]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 100
EV_PARAMS['mutation']   = point_mutation
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['100,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -55.479999999999997, 24.465682087364744, -63.149999999999999, 21.5412975468053]
CPU times: user 13min 50s, sys: 32 ms, total: 13min 51s
Wall time: 13min 51s


# Summary

In [13]:
import pandas as pd
df = pd.DataFrame.from_dict(all_measurements, orient='index')
df.columns = ['best fitness', 'mean of last gen', 'std of last gen', 'mean of best individual', 'std of best indvidiual']
df

best fitness  mean of last gen  std of last gen  \
10,false               -89            -23.40        18.558017   
50,false               -89            -47.32        28.033865   
100,false              -89            -56.41        25.786080   
10,false,5             -89            -21.73        24.185059   
50,false,5             -89            -54.25        31.271513   
100,false,5            -89            -52.43        27.603353   
10,false,10            -89            -19.93        20.559307   
50,false,10            -89            -43.21        30.966529   
100,false,10           -89            -55.48        24.465682   

              mean of best individual  std of best indvidiual  
10,false                       -36.40               18.575252  
50,false                       -53.95               26.736632  
100,false                      -63.15               23.866870  
10,false,5                     -33.50               28.355775  
50,false,5                     -61.85               28.009418  
100,false,5                    -56.35               25.737667  
10,false,10                    -32.15               24.413674  
50,false,10                    -48.45               29.537222  
100,false,10                   -63.15               21.541298

## Summary

** Without gems **

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    |  1min 36s| 89          | 23.4          | 18.56        | 36.4          | 18.58        |
| 50    |  5min 47s| 89          | 47.32         | 28.03        | 53.95         | 26.74        |
| 100   |  17min 3s| 89          | 56.41         | 25.79        | 63.15         | 23.87        |

** With gems, jbox 5 **

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    |  1min 59s| 89          | 21.73         | 24.19        | 33.5          | 28.36        |
| 50    |  5min 56s| 89          | 54.25         | 31.27        | 61.85         | 28.01        |
| 100   | 17min 57s| 89          | 52.43         | 27.61        | 56.35         | 25.74        |

** With gems, jbox 10 **

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    |  1min 46s| 89          | 19.93         | 20.56        | 32.15         | 24.41        |
| 50    |  5min  6s| 89          | 43.21         | 30.97        | 48.45         | 29.54        |
| 100   | 20min  8s| 89          | 55.48         | 24.47        | 63.15         | 21.54        |